libcontainer/apparmor/apparmor_linux.go

```diff
@@ -26,12 +27,19 @@ func isEnabled() bool {
 }
 
 func setProcAttr(attr, value string) error {
-	// Under AppArmor you can only change your own attr, so use /proc/self/
-	// instead of /proc/<tid>/ like libapparmor does
-	attrPath := "/proc/self/attr/apparmor/" + attr
+	// We need to lock ourselves to the current thread, to make sure that this
+	// thread stays alive while we are using /proc/thread-self. Otherwise, Go
+	// could switch our running thread and kill the original thread underneath
+	// us at any point where we might yield to other goroutines.
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	// Under AppArmor you can only change your own attr, so use
+	// /proc/thread-self/ instead of /proc/<tid>/ like libapparmor does.
+	attrPath := utils.ProcThreadSelf("attr/apparmor/" + attr)
```

Comment:
We still need the locking for old kernels because using `/proc/self/task/<tid>` directly has the same issues in Go as thread-self.

I agree that a release function would be nicer, and maybe we can have a "raw" version for the handful of cases where thread locking is unnecessary.

Reply:
Actually, maybe the best approach would be to do something like WithProcFd, where you pass a closure?

Reply:
I went with the closer (returned defer func) style.